### PR TITLE
Update arrow 40

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #.

# Rationale for this change

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/physical-expr/src/expressions/cast.rs [hunk 1]
```diff
@@ -33,7 +32,12 @@ use datafusion_common::{DataFusionError, Result};
 use datafusion_expr::ColumnarValue;
 
 /// provide DataFusion default cast options
-pub const DEFAULT_DATAFUSION_CAST_OPTIONS: CastOptions = CastOptions { safe: false };
```
Comment (tustvold) [inline datafusion/physical-expr/src/expressions/cast.rs:36]: This change is necessary because FormatOptions isn't const constructible, it is somewhat an accident that CastOptions has historically been - any non-trivial additional options would have broken this

File: datafusion/core/tests/sqllogictests/test_files/set_variable.slt [hunk 2]
```diff
@@ -193,31 +193,31 @@ SET TIME ZONE = '+08:00'
 query P
 SELECT '2000-01-01T00:00:00'::TIMESTAMP::TIMESTAMPTZ
 ----
-2000-01-01T08:00:00+08:00
+1999-12-31T16:00:00+08:00
```
Comment (tustvold) [inline datafusion/core/tests/sqllogictests/test_files/set_variable.slt:196]: This is the result of https://github.com/apache/arrow-rs/pull/4201

Although I'm not entirely sure it is correct... :thinking:

Comment (tustvold) [reply datafusion/core/tests/sqllogictests/test_files/set_variable.slt:196]: Ok I am fairly certain this is a bug in the scalar value logic

```
❯ create table foo (ts text) as values ('2000-01-01T00:00:00');

❯ select * from foo;
+---------------------+
| ts                  |
+---------------------+
| 2000-01-01T00:00:00 |
+---------------------+

# This is correct
❯ select arrow_cast(ts, 'Timestamp(Nanosecond, Some( "+08:00" ))') from foo;
+---------------------------+
| foo.ts                    |
+---------------------------+
| 2000-01-01T00:00:00+08:00 |
+---------------------------+

# This is incorrect
❯ select arrow_cast('2000-01-01T00:00:00', 'Timestamp(Nanosecond, Some( "+08:00" ))') from foo;
+-----------------------------+
| Utf8("2000-01-01T00:00:00") |
+-----------------------------+
| 1999-12-31T16:00:00+08:00   |
+-----------------------------+
1 row in set. Query took 0.002 seconds.
```

Comment (alamb) [reply datafusion/core/tests/sqllogictests/test_files/set_variable.slt:196]: Yes, I think a bunch of that logic is replicated in scalar.rs (as it wasn't available upstream in arrow-rs)

I think @berkaysynnada  may have some thoughts / be thinking about this.

File: datafusion/common/src/scalar.rs [hunk 3]
```diff
@@ -1760,17 +1760,17 @@ macro_rules! build_array_from_option {
             None => new_null_array(&DataType::$DATA_TYPE($ENUM), $SIZE),
         }
     }};
-    ($DATA_TYPE:ident, $ENUM:expr, $ENUM2:expr, $ARRAY_TYPE:ident, $EXPR:expr, $SIZE:expr) => {{
+}
+
+macro_rules! build_timestamp_array_from_option {
```
Comment (tustvold) [inline datafusion/common/src/scalar.rs:1765]: I split this out into its own macro as overloading based on number of arguments is kind of hard to parse

File: datafusion/common/src/scalar.rs [hunk 4]
```diff
@@ -1760,17 +1760,17 @@ macro_rules! build_array_from_option {
...
+    ($TIME_UNIT:expr, $TZ:expr, $ARRAY_TYPE:ident, $EXPR:expr, $SIZE:expr) => {
         match $EXPR {
             Some(value) => {
-                let array: ArrayRef = Arc::new($ARRAY_TYPE::from_value(*value, $SIZE));
-                // Need to call cast to cast to final data type with timezone/extra param
-                cast(&array, &DataType::$DATA_TYPE($ENUM, $ENUM2))
-                    .expect("cannot do temporal cast")
+                Arc::new($ARRAY_TYPE::from_value(*value, $SIZE).with_timezone_opt($TZ))
```
Comment (tustvold) [inline datafusion/common/src/scalar.rs:1769]: This is the major change, we use with_timezone_opt instead of cast, as cast will now actually perform timezone conversion

File: datafusion/core/src/physical_plan/joins/symmetric_hash_join.rs [hunk 5]
```diff
@@ -458,7 +458,7 @@ impl ExecutionPlan for SymmetricHashJoinExec {
             DisplayFormatType::Default => {
                 let display_filter = self.filter.as_ref().map_or_else(
                     || "".to_string(),
-                    |f| format!(", filter={:?}", f.expression()),
+                    |f| format!(", filter={}", f.expression()),
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/joins/symmetric_hash_join.rs:461]: This change is so that we don't output all of CastOptions and by extension FormatOptions. It is also more correct to not be printing a debug representation as part of Display

File: datafusion/core/tests/sqllogictests/test_files/arrow_typeof.slt [hunk 6]
```diff
@@ -321,13 +321,12 @@ select arrow_cast(timestamp '2000-01-01T00:00:00', 'Timestamp(Nanosecond, Some(
 query P
 select arrow_cast(timestamp '2000-01-01T00:00:00', 'Timestamp(Nanosecond, Some( "+08:00" ))');
 ----
-2000-01-01T08:00:00+08:00
+2000-01-01T00:00:00+08:00
```
Comment (tustvold) [inline datafusion/core/tests/sqllogictests/test_files/arrow_typeof.slt:324]: :tada: It's correct now

**Review by jackwener — APPROVED** [review]

**Comment by jackwener** [comment]
Thanks @tustvold !❤️
